sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -3574,18 +3574,29 @@ __ESIMD_API simd<T, N> slm_atomic_update_impl(simd<uint32_t, N> offsets,
 template <atomic_op Op, typename T, int N, class Tx = detail::__raw_t<T>>
 __ESIMD_API std::enable_if_t<__ESIMD_DNS::get_num_args<Op>() == 0, simd<T, N>>
 slm_atomic_update(simd<uint32_t, N> byte_offset, simd_mask<N> mask = 1) {
-  // uint16_t and load are supported only by LSC.
-  if constexpr (sizeof(T) == 2 || Op == atomic_op::load ||
-                !__ESIMD_DNS::isPowerOf2(N, 32)) {
-    return slm_atomic_update_impl<Op, T, N,
-                                  detail::lsc_data_size::default_size>(
-        byte_offset, mask);
+  if constexpr (Op == atomic_op::load) {
```

Comment:
This indeed enables support of atomic_op::load operation for Gen12, but the order of 'if' needs to be different.
As it is written now this code creates the emulation sequence as the first move, i.e. it does it even for PVC, which has native support of op::load.

Reply:
Not sure I follow, we cannot statically know if it is gen12 or pvc. No combination of ifs can resolve that, right?

Reply:
Hmm, you're right. We still can avoid emulation for some cases.
```
if constexpr (sizeof(T) == 2 || !__ESIMD_DNS::isPowerOf2(N, 32)) {
  call LSC code, NO emulation for load
} else if (atomic_op::load) {
  emulation for load
} else {
  Gen12/dword atomic call
}
```

Reply:
Done.